### PR TITLE
addUser & getUserById methods functionality

### PR DESCRIPTION
**Pull request by Plymatea**
**1: Ticket**
None

**2: What this does:**

- Converts Class User to Interface IUser
- Adds UserAPI property "nextId"
- Add UserAPI private createUser()
- Privatizes UserAPI properties
- Adds functionality to UserAPI addUser() && getUserById()
- Throws error when user submits data with pre-existing ID field
- Throws error when user submits duplicate user to be entered into database. *(checks for equality across name, favcolor, age properties)*

**3: Media**

Code to: 
- create API
- pass user into addUser()
- find user by getUserById()
image- Terminal window showing console.log of code above
image

**4: Test**
Compiles correctly with only expected errors present.

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -1,32 +1,37 @@
...
   private assignId() {
-
+    this.nextId++
+    return this.nextId
   }
 
-  addUser(user: User) {
-
+  addUser(user: IUser) {
+    const {name, favColor, age} = user
```
Comment (rachoac) [inline src/index.ts]: if the user has an id already, this function should throw an exception (see javascript.info for how to throw errors)

File: src/index.ts [hunk 2]
```diff
@@ -1,32 +1,37 @@
...
+    return newUser
   }
 
   getUserById(id: number) {
-
+    const foundUser = this.users.find( user => user.id === id) 
+    if (foundUser) {
+      return foundUser
+    }
+    return "No user found"
```
Comment (rachoac) [inline src/index.ts]: throw an error instead of returning a string or an object

File: src/index.ts (hunk 1 above)
Comment (Plymatea) [reply src/index.ts]: Can I get some clarifications here: 
1) Will the user object being passed to the API have a pre-existing ID property that should be stored in the database?
1a) If so, the error checking is to verify no two users have the same ID, correct? 
2) You want error checking for duplicate users. ex name: {"andy" age: 111}  => Error: User already exists in the database

Comment (rachoac) [reply src/index.ts]: `Will the user object being passed to the API have a pre-existing ID property that should be stored in the database?` someone calling the addUser method could formulate a user object with the id already set. This is an error condition, becuase we want the userapi to be the one assigning ids.

`You want error checking for duplicate users. ex name: {"andy" age: 111} => Error: User already exists in the database` correct

Comment (Plymatea) [reply src/index.ts]: Thank you

File: src/index.ts [hunk 3]
```diff
@@ -1,32 +1,53 @@
...
-
+    try {
+      const foundUser = this.users.find( user => user.id === id) 
+      if (foundUser) {
+        return foundUser
+      }
+      throw "Error: No user found."
+    }
+    catch (error) {
+      return (error)
```
Comment (rachoac) [inline src/index.ts]: we should not be returning errors ... let the function throw the error and have its caller deal with it

File: src/index.ts [hunk 4]
```diff
@@ -1,32 +1,53 @@
...
-
+  addUser(user: IUser) {
+    const result = (this.users).find( ( { name, favColor, age }: IUser) => (name === user.name && favColor === user.favColor && age === user.age))
+    try {
+      if (user.id) {
+        throw "Error: User Id is set by database. Please resubmit without pre-existing value."
+      } else if (result) {
+        throw "Error: A user with those properties already exists in the database."
+      } 
+    } catch (error) {
```
Comment (rachoac) [inline src/index.ts]: see comment below

File: src/index.ts (hunk 3 above)
Comment (rachoac) [reply src/index.ts]: be sure to have a case in the console logs below that intentionalyl induces the error and catches for it there

**Pull request merged**